crates/xtask/src/changelog.rs

```diff
@@ -0,0 +1,549 @@
+    Patch,
+}
+
+#[derive(Debug, Default, PartialEq, Eq)]
+struct Changelog {
+    releases: Vec<Release>,
+
+    /// Incremented to acknowledge the changelog is already up-to-date.
+    ///
+    /// There is a CI test to make sure the changelog is updated when the crate is modified. Some
+    /// modifications are already documented in the changelog, so the test will have a false
+    /// positive. This counter is used to artificially modify the changelog and explicitly
+    /// acknowledge that it's already up-to-date.
+    skip_counter: u32,
+}
+
+impl Changelog {
+    fn read_file(changelog_path: &str) -> Result<Changelog> {
+        let changelog_contents = wasefire_cli_tools::fs::read(&changelog_path)?;
+
+        Self::parse(String::from_utf8(changelog_contents)?)
+    }
+
+    /// Converts raw file contents into a Changelog data structure.
+    ///
+    /// Validates requirements in docs/contributing/changelog.md
+    fn parse(contents: String) -> Result<Changelog> {
+        let mut skip_counter = 0;
+        let mut releases: Vec<Release> = Vec::new();
+
+        let mut lines = contents.lines().filter(|line| !line.is_empty()).peekable();
+
+        // First line is H1.
+        if lines.next() != Some("# Changelog") {
+            return Err(anyhow!("H1 with 'Changelog' is required"));
+        }
+
+        while let Some(mut current_line) = lines.next() {
+            let mut next_line = lines.peek();
+
+            // Parse a whole release
+            if let Some(version_string) = current_line.strip_prefix("## ") {
+                let mut release = Release::from(version_string)?;
+
+                // Parse each release type (major, minor, patch)
+                while next_line.is_some_and(|line| line.starts_with("### ")) {
+                    // Advance
+                    current_line = lines.next().unwrap();
+                    next_line = lines.peek();
+
+                    let release_type_string = current_line.trim_start_matches("### ");
+                    let release_type = ReleaseType::from_str(release_type_string, true)
+                        .map_err(|err| anyhow!("Failed to parse version_type: {err}"))?;
+
+                    let mut release_descriptions = Vec::new();
+
+                    // Parse the release descriptions
+                    // Some lines may be a new description. Some lines may be continuation of
+                    // previous descriptions.
+                    while next_line
+                        .is_some_and(|line| line.starts_with("- ") || line.starts_with("  "))
+                    {
+                        // Advance
+                        current_line = lines.next().unwrap();
+                        next_line = lines.peek();
+
+                        // New description
+                        if current_line.starts_with("- ") {
+                            release_descriptions
+                                .push(current_line.trim_start_matches("- ").to_string());
```

Comment:
I would avoid using `trim`. Let's use `strip_prefix().unwrap()` if needed, otherwise ideally we would `match strip_prefix()` and handle both cases, but I can see how it's not always convenient.